kernel/objects/capability-spinning/objects/RevokeOperation.cc

```diff
@@ -99,98 +99,92 @@ namespace mythos {
+      if (leafCap.getPtr() == _guarded) {
+        leaf->unlock_cap();
+        // attempted to delete guarded object
+        THROW(Error::CYCLIC_DEPENDENCY);
+      }
+      leaf->lock_cap();
+      if (leaf->cap() != leafCap) {
+        MLOG_DETAIL(mlog::cap, "leaf cap changed concurrently");
+        leaf->unlock_cap();
+        continue;
+      }
+      auto delRes = leafCap.getPtr()->deleteCap(*leaf, leafCap, *this);
+      auto prevres = leaf->lock_prev();
+      ASSERT_MSG(prevres, "somebody unlinked CapEntry currently in unlinking process");
+      leaf->lock_next();
+      if (delRes) {
+        leaf->unlinkAndUnlockLinks();
+        leaf->reset();
+      } else {
+        // deletion failed
+        // Either tried to delete a portal that is currently deleting
+        // or tried to to delete _guarded via a recursive call.
+        leaf->unlock_cap();
+        RETHROW(delRes);
+      }
     } while (leaf != root);
     // deleted root
     RETURN(Error::SUCCESS);
   }
 
-  bool RevokeOperation::_startTraversal(CapEntry* root, Cap rootCap)
-  {
-    if (!root->lock_prev()) {
-      // start is currently unlinked
-      // must be the work of another deleter ... success!
-      return false;
-    }
-    if (root->prev() == root) {
-      // this is the actual root of the tree
-      // and has no children
-      // avoid deadlocks and finish the operation
-      root->unlock_prev();
-      return false;
-    }
-    root->lock();
-    // compare only value, not the zombie state
-    if (root->cap().asZombie() != rootCap.asZombie()) {
-      // start has a new value
-      // must be the work of another deleter ... success!
-      root->unlock();
-      root->unlock_prev();
-      return false;
-    }
-    return true;
-  }
-
-  CapEntry* RevokeOperation::_findLockedLeaf(CapEntry* root)
+  // not sure if we even need that locking
+  bool RevokeOperation::_findLeaf(CapEntry* const root, Cap const rootCap, CapEntry*& leafEntry, Cap& leafCap)
   {
-    auto curEntry = root;
+    MLOG_ERROR(mlog::cap, __PRETTY_FUNCTION__, DVAR(root) );
+    leafEntry = root;
+    leafCap = rootCap;
     while (true) {
-      auto curCap = curEntry->cap();
-      auto nextEntry = curEntry->next();
-      // wait for potencially allocated cap to become usable/zombie
-      Cap nextCap;
-      for (nextCap = nextEntry->cap();
-          nextCap.isAllocated();
-          nextCap = nextEntry->cap()) {
-        ASSERT(!nextEntry->isDeleted());
-        hwthread_pause();
+      auto nextEntry = leafEntry->next();
+      if (nextEntry) {
+        Cap nextCap;
+        // wait for potencially allocated cap to become usable/zombie
+        for (nextCap = nextEntry->cap();
+            nextCap.isAllocated();
+            nextCap = nextEntry->cap()) {
+          ASSERT(!nextEntry->isDeleted());
+          hwthread_pause();
+        }
+        if (cap::isParentOf(*leafEntry, leafCap, *nextEntry, nextCap)) {
+          if (!nextEntry->kill(nextCap)) {
```

Comment:
modifying anything in the chain from within a function named `findLeaf` feels very wrong. This line would kill every CapEntry on the way to the first leaf. Why is this necessary? Or what is the benefit? Why does this no permanent damage in the case that the deletion operation has to abort?

Reply:
I rather change the name. Killing the intermediate nodes ensures they can't get inherited from and this subtree can't grow further while deleting.

Killing the cap should not do damage, it can be deleted by concurrent operations as well as in a _retry_.